### PR TITLE
ALTAPPS-498: Shared local logging

### PR DESCRIPTION
**Pull request by XanderZhu**
**YouTrack Issues**:
[ALTAPPS-498](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-498)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] All checks have been passed;
- [x] Changes have been checked locally.

**Description**
- Add multiplatform logger [Kermit](https://kermit.touchlab.co/)
- Add LoggerCompoment to user predefined logger in the app
- Add SentryLogWriter to write logs into Sentry
- Add StateReducer.wrapWithLogger() extension to log state and actions procured by reducer

### CODE REVIEW

File: shared/src/commonMain/kotlin/org/hyperskill/app/logging/inject/LoggerComponentImpl.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package org.hyperskill.app.logging.inject
+
+import co.touchlab.kermit.Logger
+import org.hyperskill.app.core.injection.AppGraph
+
+class LoggerComponentImpl(private val appGraph: AppGraph) : LoggerComponent {
+    override val logger: Logger
+        get() = LoggerBuilder.build(
```
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/logging/inject/LoggerComponentImpl.kt:8]: Oops, my fault.

File: shared/src/commonMain/kotlin/org/hyperskill/app/logging/inject/LoggerBuilder.kt
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/logging/inject/LoggerBuilder.kt:1]: platformLogWriter is an `expect` factory function for platform-specific log writers.
https://github.com/touchlab/Kermit/blob/main/kermit-core/src/commonMain/kotlin/co/touchlab/kermit/platformLogWriter.kt
The default iOS implementation is [XcodeSeverityWriter](https://github.com/touchlab/Kermit/blob/main/kermit-core/src/darwinMain/kotlin/co/touchlab/kermit/XcodeSeverityWriter.kt)

File: shared/src/commonMain/kotlin/org/hyperskill/app/logging/SentryLogWriter.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            Severity.Assert -> false
+            Severity.Info,
+            Severity.Warn,
+            Severity.Error -> true
+        }
+
+    override fun log(severity: Severity, message: String, tag: String, throwable: Throwable?) {
+        val formattedMessage = messageStringFormatter.formatMessage(null, Tag(tag), Message(message))
+        when (severity) {
+            Severity.Info -> sentryInteractor.captureMessage(formattedMessage, HyperskillSentryLevel.INFO)
```
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/logging/SentryLogWriter.kt:30]: This is how feature state reducing looks in sentry console
https://sentry.int.hyperskill.org/organizations/hyperskill/issues/267001/?project=4&query=is%3Aunresolved+level%3Adebug&referrer=issue-stream&statsPeriod=14d
Similar to errors.

File: shared/src/commonMain/kotlin/org/hyperskill/app/logging/SentryLogWriter.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    override fun isLoggable(tag: String, severity: Severity): Boolean =
+        when (severity) {
+            Severity.Verbose,
+            Severity.Debug,
+            Severity.Assert -> false
+            Severity.Info,
+            Severity.Warn,
+            Severity.Error -> true
+        }
```
[the remark this answers is not in the document]
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/logging/SentryLogWriter.kt:25]: @XanderZhu OK, I understood - Sentry creates issues for messages.

We can also think about this in the future - we can use Sentry breadcrumbs for debugging features state changes (like analytic events breadcrumbs), which might be helpful for debugging crashes.

Screenshot 2023-09-29 at 10 33 14 AM

File: shared/src/commonMain/kotlin/org/hyperskill/app/logging/SentryLogWriter.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/logging/SentryLogWriter.kt:30]: > This is how feature state reducing looks in the sentry console (Similar to errors.)

Am I correct, we will not log state reducing to the Sentry (`wrapWithLogger / LoggableStateReducer`) at all?
according to your [comment](https://github.com/hyperskill/mobile-app/pull/682#discussion_r1340218691) I also think it's useless, I think it's better to think about [breadcrumbs](https://github.com/hyperskill/mobile-app/pull/682#discussion_r1340951402)

**Review by ivan-magda — APPROVED** [review]

**Comment by XanderZhu** [comment]
@vladkash 
Do you have any comments on this issue?

File: shared/src/commonMain/kotlin/org/hyperskill/app/logging/SentryLogWriter.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/logging/SentryLogWriter.kt:30]: @ivan-magda 
By default, state changes will not be logged into sentry.
But it depends on the severity. At the moment minSeverity for SentryLogWritter is `Info` and the default severity for `wrapWithLogger` is `Debug`.

[the remark this answers is not in the document]
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/logging/SentryLogWriter.kt:30]: OK

**Comment by vladkash** [comment]
@XanderZhu no, I don't have them

**Pull request merged**